crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+#[derive(Clone, Debug)]
+enum ModuleTypeDef {
+    TypeDef(crate::core::Type),
+    Import(crate::core::Import),
+    Export(String, crate::core::EntityType),
+}
+
+#[derive(Clone, Debug)]
+enum Alias {
+    InstanceExport {
+        instance: u32,
+        name: String,
+        kind: InstanceExportAliasKind,
+    },
+    Outer {
+        count: u32,
+        i: u32,
+        kind: OuterAliasKind,
+    },
+}
+
+#[derive(Clone, Debug)]
+enum InstanceExportAliasKind {
+    Module,
+    Component,
+    Instance,
+    Func,
+    Value,
+    Table,
+    Memory,
+    Global,
+}
+
+#[derive(Clone, Debug)]
+enum OuterAliasKind {
+    Module,
+    Component,
+    Type,
+}
+
+#[derive(Clone, Debug)]
+struct ComponentType {
+    defs: Vec<ComponentTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum ComponentTypeDef {
+    Instance(InstanceTypeDef),
+    Import(Import),
+}
+
+#[derive(Clone, Debug)]
+struct InstanceType {
+    defs: Vec<InstanceTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum InstanceTypeDef {
+    Type(Type),
+    Export { name: String, ty: u32 },
+    Alias(Alias),
+}
+
+#[derive(Clone, Debug)]
+struct FuncType {
+    params: Vec<NamedType>,
+    result: ValueType,
+}
+
+#[derive(Clone, Debug)]
+struct NamedType {
+    name: String,
+    ty: ValueType,
+}
+
+#[derive(Clone, Debug)]
+enum ValueType {
+    Indexed(u32),
+    Unit,
+    Bool,
+    S8,
+    U8,
+    S16,
+    U16,
+    S32,
+    U32,
+    S64,
+    U64,
+    F32,
+    F64,
+    Char,
+    String,
+}
+
+#[derive(Clone, Debug)]
+enum CompoundType {
+    Record(RecordType),
+    Variant(VariantType),
+    List(ListType),
+    Tuple(TupleType),
+    Flags(FlagsType),
+    Enum(EnumType),
+    Union(UnionType),
+    Optional(OptionalType),
+    Expected(ExpectedType),
+    Named(NamedType),
+}
+
+#[derive(Clone, Debug)]
+struct RecordType {
+    fields: Vec<NamedType>,
+}
+
+#[derive(Clone, Debug)]
+struct VariantType {
+    cases: Vec<NamedType>,
+    default: Option<u32>,
+}
+
+#[derive(Clone, Debug)]
+struct ListType {
+    elem_ty: ValueType,
+}
+
+#[derive(Clone, Debug)]
+struct TupleType {
+    fields: Vec<ValueType>,
+}
+
+#[derive(Clone, Debug)]
+struct FlagsType {
+    fields: Vec<String>,
+}
+
+#[derive(Clone, Debug)]
+struct EnumType {
+    variants: Vec<String>,
+}
+
+#[derive(Clone, Debug)]
+struct UnionType {
+    variants: Vec<String>,
+}
+
+#[derive(Clone, Debug)]
+struct OptionalType {
+    inner_ty: ValueType,
+}
+
+#[derive(Clone, Debug)]
+struct ExpectedType {
+    ok_ty: ValueType,
+    err_ty: ValueType,
+}
+
+#[derive(Debug)]
+struct ImportSection {
+    imports: Vec<Import>,
+}
+
+#[derive(Clone, Debug)]
+struct Import {
+    name: String,
+    ty: u32,
+}
+
+#[derive(Debug)]
+struct FuncSection {}
+
+#[derive(Debug)]
+struct CoreSection {}
+
+#[derive(Debug)]
+struct ComponentSection {}
+
+#[derive(Debug)]
+struct InstanceSection {}
+
+#[derive(Debug)]
+struct ExportSection {}
+
+#[derive(Debug)]
+struct StartSection {}
+
+#[derive(Debug)]
+struct AliasSection {}
```

Comment:
FWIW I think it's fine to remove these empty structs and the empty methods with `todo!()` for now and we can add them back in as they get implemented

Reply:
This would just be churn on my end. I'll have to add them back as soon as I merge this skeleton PR, and this PR isn't going into `main`, so I figured I could leave them as placeholders for now.